### PR TITLE
[Recipe PR] MELD: Multimodal EmotionLines Dataset

### PR DESCRIPTION
**Pull request by realzza**
### About
Added baseline recipe for MELD: Multimodal EmotionLines Dataset

### Checklist
- [x] (meld) be careful about the name for the recipe. It is recommended to follow naming conventions of the other recipes
- [x] common/shared files are linked with **soft link** (see Section 1.3.3)
- [x] modified or new python scripts should be passed through **latest** black formating (by using python package black). The command to be executed could be `black espnet espnet2 test utils setup.py egs*/*/*/local egs2/TEMPLATE/*/pyscripts tools/*.py ci/*.py`
- [x] modified or new python scripts should be passed through **latest** isort formating (by using python package isort). The command to be executed could be `isort espnet espnet2 test utils setup.py egs*/*/*/local egs2/TEMPLATE/*/pyscripts tools/*.py ci/*.py`
- [x] cluster settings should be set as **default** (e.g., cmd.sh conf/slurm.conf conf/queue.conf conf/pbs.conf)
- [x] update `egs/README.md` or `egs2/README.md` with corresponding recipes
- [x] add corresponding entry in `egs2/TEMPLATE/db.sh` for a new corpus
- [x] try to **simplify** the model configurations. We recommend to have only the best configuration for the start of a recipe. Please also follow the default rule defined in Section 1.3.3
- [x] large meta-information for a corpus should be maintained elsewhere other than in the recipe itself
- [x] recommend to also include results and pre-trained model with the recipe
    - [x] results
    - [x] (TODO) model

### Todo
- [x] submit model through huggingface

### CODE REVIEW

**Comment by sw005320** [comment]
@siddhu001, it would be great if you also review this PR.

File: egs2/meld/asr1/local/data_prep.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            .lower()
+            .split()
+        )
+        utt2spk.append(
+            f"{spk}-dia{line[5]}-utt{line[6]}-sea{line[7]}-epi{line[8]}-{phase} {spk}"
+        )
+        text.append(
+            f"{spk}-dia{line[5]}-utt{line[6]}-sea{line[7]}-epi{line[8]}-{phase} {line[3]} {curr_line}"
+        )
+        # ffmpeg -i /ocean/projects/cis210027p/zzhou5/data/meld/MELD-RAW/MELD.Raw/dev/dev_splits_complete/dia105_utt9.mp4 -ac 1 -ar 16000 -vn /ocean/projects/cis210027p/zzhou5/data/meld/MELD-RAW/meld.audio/dev2/dia105_utt9.wav
```
Comment (ftshijt) [inline egs2/meld/asr1/local/data_prep.py:63]: you can just remove lines that commented out

File: egs2/meld/asr1/local/data_prep.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        wav_scp.append(
+            f"{spk}-dia{line[5]}-utt{line[6]}-sea{line[7]}-epi{line[8]}-{phase} ffmpeg -i {meld_root}/wavs/{phase}/dia{line[5]}_utt{line[6]}.mp4 -ac 1 -ar 16000 -f wav -vn -hide_banner -loglevel error - |"
+        )
+        # wav_scp.append(f'{spk}-dia{line[5]}-utt{line[6]}-sea{line[7]}-epi{line[8]}-{phase} {meld_root}/wavs/{phase}/dia{line[5]}_utt{line[6]}.wav')
```
Comment (ftshijt) [inline egs2/meld/asr1/local/data_prep.py:67]: Ditto

File: egs2/meld/asr1/local/path.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+if ! which ffmpeg &> /dev/null; then
```
Comment (ftshijt) [inline egs2/meld/asr1/local/path.sh:1]: I remember the ffmpeg is automatically installed during the time of torchaudio installation? Do we need this check

File: egs2/meld/asr1/run.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+train_set="train"
+valid_set="valid"
+test_sets="valid test"
+
+asr_config=conf/tuning/train_asr_hubert_transformer_adam_specaug_meld.yaml
```
Comment (ftshijt) [inline egs2/meld/asr1/run.sh:12]: Please use a soft link and make this as the `conf/train_asr.yaml` for consistency to other recipes.

File: egs2/meld/asr1/run.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+inference_config=conf/decoder_asr.yaml
+
+./asr.sh \
+    --lang en \
+    --ngpu 1 \
+    --nj 8 \
+    --stop_stage 13 \
```
Comment (ftshijt) [inline egs2/meld/asr1/run.sh:19]:
```suggestion
```

File: egs2/meld/asr1/run.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    --inference_config "${inference_config}" \
+    --use_lm false \
+    --nbpe 850 \
+    --token_type bpe\
+    --bpe_train_text "dump/raw/train/text" \
```
Comment (ftshijt) [inline egs2/meld/asr1/run.sh:24]: This is automatically set, we can safely remove this.

File: egs2/meld/asr1/run.sh [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+valid_set="valid"
+test_sets="valid test"
+
+asr_config=conf/tuning/train_asr_hubert_transformer_adam_specaug_meld.yaml
+inference_config=conf/decoder_asr.yaml
+
+./asr.sh \
+    --lang en \
+    --ngpu 1 \
+    --nj 8 \
```
Comment (ftshijt) [inline egs2/meld/asr1/run.sh:18]:
```suggestion
```

File: egs2/meld/asr1/run.sh [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    --stop_stage 13 \
+    --inference_config "${inference_config}" \
+    --use_lm false \
+    --nbpe 850 \
+    --token_type bpe\
+    --bpe_train_text "dump/raw/train/text" \
+    --bpe_nlsyms sadness,surprise,neutral,joy,anger,fear,disgust\
+    --audio_format wav\
+    --feats_type raw\
```
Comment (ftshijt) [inline egs2/meld/asr1/run.sh:27]:
```suggestion
```
Since they are default, we do not need the speicification

File: egs2/meld/asr1/run.sh [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    --max_wav_duration 20 \
+    --feats_normalize utterance_mvn\
+    --inference_nj 4 \
+    --inference_asr_model valid.acc.ave_5best.pth\
```
Comment (ftshijt) [inline egs2/meld/asr1/run.sh:31]:
```suggestion
```

File: egs2/meld/asr1/README.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
+# MELD RESULTS
+## Environments
+- date: `Thu Nov 10 09:07:40 EST 2022`
+- python version: `3.8.6 (default, Dec 17 2020, 16:57:01)  [GCC 10.2.0]`
+- espnet version: `espnet 202207`
+- pytorch version: `pytorch 1.8.1+cu102`
+- Git hash: `a7bd6522b32ec6472c13f6a2289dcdff4a846c12`
+  - Commit date: `Wed Sep 14 08:34:27 2022 -0400`
```
Comment (ftshijt) [inline egs2/meld/asr1/README.md:8]: Please also try to add a model for huggingface

**Review by ftshijt — CHANGES_REQUESTED** [review]
Just to confirm: it seems that this is only an unimodal model. Is it intentional or you will keep update it.

I also recommend you to check the task of SLU, which could be a better place to hold your recipe as well (we can keep on this).

File: egs2/meld/asr1/local/path.sh (hunk 3 above)
Comment (realzza) [reply egs2/meld/asr1/local/path.sh:1]: This check helped me discover `ffmpeg` was missing in my environment as I re-run this recipe. Found something [here](https://pytorch.org/audio/master/tutorials/streamwriter_basic_tutorial.html#:~:text=TorchAudio%20dynamically%20loads%20compatible%20FFmpeg%20libraries%20installed%20on%20the%20system.), seems that torchaudio will automatically load ffmpeg, given ffmpeg is installed. If that's the case, we may need this check.

File: egs2/meld/asr1/run.sh (hunk 4 above)
Comment (realzza) [reply egs2/meld/asr1/run.sh:12]: If I understand you right, do you suggest softlink a `conf/train_asr.yaml` to `conf/`, and replace line12 with `asr_config=conf/train_asr.yaml`, or just add the soft link and keep the tuning configs? Some recipes use tuning config in their `run.sh`.

Comment (siddhu001) [reply egs2/meld/asr1/run.sh:12]: softlink a conf/train_asr.yaml to your actual config(conf/,) and replace line12 with asr_config=conf/train_asr.yaml

File: egs2/meld/asr1/local/data_prep.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+            f"{spk}-dia{line[5]}-utt{line[6]}-sea{line[7]}-epi{line[8]}-{phase} ffmpeg -i {meld_root}/wavs/{phase}/dia{line[5]}_utt{line[6]}.mp4 -ac 1 -ar 16000 -f wav -vn -hide_banner -loglevel error - |"
+        )
+
+    with open(os.path.join("data", phase, "utt2spk"), "w") as f:
+        f.write("\n".join(utt2spk))
+    with open(os.path.join("data", phase, "text"), "w") as f:
+        f.write("\n".join(text))
+    with open(os.path.join("data", phase, "wav.scp"), "w") as f:
+        f.write("\n".join(wav_scp))
+    with open(os.path.join("data", phase, "transcript"), "w") as f:
```
Comment (siddhu001) [inline egs2/meld/asr1/local/data_prep.py:73]: You do not need to create transcript file. It is not needed.

**Comment by siddhu001** [comment]
I would suggest incorporating all of Jiatong's feedback. I added a comment and clarification, rest looks good to me.

File: egs2/meld/asr1/local/path.sh (hunk 3 above)
Comment (ftshijt) [reply egs2/meld/asr1/local/path.sh:1]: I see. I got it installed when using conda, but maynot with pip. So we can just keep it~

File: egs2/meld/asr1/run.sh (hunk 6 above)
Comment (realzza) [reply egs2/meld/asr1/run.sh:24]: Thanks, I will remove `--use_lm` and `--bpe_train_text`. I'm not sure if `--nbpe` and `--token_type` are set by default.

File: egs2/meld/asr1/README.md (hunk 10 above)
Comment (realzza) [reply egs2/meld/asr1/README.md:8]: Just made a request to join `huggingface/espnet`, should be able to add the model soon.

Comment (realzza) [reply egs2/meld/asr1/README.md:8]: model submitted [here](https://huggingface.co/espnet/realzza-meld-asr-hubert-transformer)

Comment (sw005320) [reply egs2/meld/asr1/README.md:8]: You can add a link like the other recipe README.md

Comment (realzza) [reply egs2/meld/asr1/README.md:8]: Just added!

Comment (sw005320) [reply egs2/meld/asr1/README.md:8]: cool!

Comment (realzza) [reply egs2/meld/asr1/README.md:8]: Just added the HF model link!

**Comment by sw005320** [comment]
@ftshijt, can you review it again?
I think this PR is almost ready.

**Comment by ftshijt** [comment]
Looks cool! After it passed the CI tests, I will merge it

**Comment by realzza** [comment]
Seems it fails the CI test, as well as centos7 and debian9 tests due to the overlength lines in `data_prep.py`, let me fix this format issue shortly!

**Comment by realzza** [comment]
Hi @ftshijt, thank you for your review! I fixed the flake8 error `W605 invalid escape sequence '\-'`, we can rerun the CI tests.

**Comment by realzza** [comment]
The failed check may be caused by this [issue](https://github.com/numpy/numpy/issues/22623) reported in the latest version of numpy.

**Comment by sw005320** [comment]
Thanks, @realzza!

**Pull request merged**